File: lib/core/socket/protocol/message_dedup.dart
/// REFERENCE implementation of server-side message deduplication
/// (idempotency by message_id) plus the client-side mirror used by the
/// outbox when retrying after a reconnection.
///
/// Idempotency rule: a `message.send` carrying a message_id that was already
/// accepted is NOT re-persisted and NOT re-fanned-out; the server answers
/// with the ORIGINAL ack (same server_seq) so the retrying client converges.
class MessageDedup {
  MessageDedup({this.windowSize = 4096});

  /// Number of recent ids remembered (LRU). Production mapping: Redis SET
  /// with TTL per account/device.
  final int windowSize;

  final Map<String, int> _seen = <String, int>{};
  int _clockTick = 0;

  /// Returns true if [messageId] is new (and remembers it).
  /// Returns false for a duplicate.
  bool accept(String messageId) {
    if (_seen.containsKey(messageId)) {
      _seen[messageId] = _clockTick++;
      return false;
    }
    _seen[messageId] = _clockTick++;
    if (_seen.length > windowSize) {
      _evictOldest();
    }
    return true;
  }

  /// True when the id is already known (without consuming).
  bool contains(String messageId) => _seen.containsKey(messageId);

  void _evictOldest() {
    String? oldest;
    var oldestTick = 1 << 62;
    for (final entry in _seen.entries) {
      if (entry.value < oldestTick) {
        oldestTick = entry.value;
        oldest = entry.key;
      }
    }
    if (oldest != null) _seen.remove(oldest);
  }

  int get remembered => _seen.length;
}

File: server/src/protocol/message_dedup.ts
/**
 * Server-side message deduplication (idempotency by message_id).
 *
 * TypeScript port of `lib/core/socket/protocol/message_dedup.dart`.
 *
 * Idempotency rule: a `message.send` carrying a message_id that was already
 * accepted is NOT re-persisted and NOT re-fanned-out; the server answers
 * with the ORIGINAL ack (same server_seq) so the retrying client converges.
 *
 * LRU window in process memory; production mapping: Redis `SET NX` with
 * 24h TTL per account (docs/SOCKET_SERVER_ARCHITECTURE.md §4).
 */
export class MessageDedup {
  readonly windowSize: number;
  private readonly seen = new Map<string, number>();
  private clockTick = 0;

  constructor(windowSize = 4096) {
    this.windowSize = windowSize;
  }

  /** Returns true if messageId is new (and remembers it). */
  accept(messageId: string): boolean {
    if (this.seen.has(messageId)) {
      this.seen.set(messageId, this.clockTick++);
      return false;
    }
    this.seen.set(messageId, this.clockTick++);
    if (this.seen.size > this.windowSize) this.evictOldest();
    return true;
  }

  /** True when the id is already known (without consuming). */
  contains(messageId: string): boolean {
    return this.seen.has(messageId);
  }

  private evictOldest(): void {
    let oldest: string | null = null;
    let oldestTick = Number.MAX_SAFE_INTEGER;
    for (const [id, tick] of this.seen) {
      if (tick < oldestTick) {
        oldestTick = tick;
        oldest = id;
      }
    }
    if (oldest !== null) this.seen.delete(oldest);
  }

  get remembered(): number {
    return this.seen.size;
  }
}
